### PR TITLE
Add Role V2 definitions

### PR DESCRIPTION
**Pull request by AnuradhaSK**
## Purpose
Part of https://github.com/wso2/product-is/issues/16363
Introduce definitions for RoleV2 resource and it's management endpoints

### CODE REVIEW

File: modules/charon-core/src/main/java/org/wso2/charon3/core/objects/Group.java [hunk 1]
```diff
@@ -411,4 +413,104 @@ private void setRole(ComplexAttribute groupPropertiesAttribute) throws CharonExc
...
+        String audienceDisplay = role.getAudienceDisplayName();
+        String audienceType = role.getAudienceType();
+        ComplexAttribute complexAttribute = new ComplexAttribute();
+
+        if (StringUtils.isNotBlank(value)) {
+            valueSimpleAttribute = new SimpleAttribute(SCIMConstants.CommonSchemaConstants.VALUE, value);
+            valueSimpleAttribute = (SimpleAttribute) DefaultAttributeFactory
+                    .createAttribute(SCIMSchemaDefinitions.SCIMGroupSchemaDefinition.ROLES_VALUE, valueSimpleAttribute);
+            complexAttribute.setSubAttribute(valueSimpleAttribute);
+        }
```
Comment (thanujalk) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/objects/Group.java:444]:
```suggestion
        if (StringUtils.isNotBlank(value)) {
            complexAttribute.setSubAttribute(getSubAttribute(SCIMConstants.CommonSchemaConstants.VALUE, SCIMSchemaDefinitions.SCIMGroupSchemaDefinition.ROLES_VALUE, value));
        }
        
        private SimpleAttribute getSubAttribute(String a, String b, String c) {
           SimpleAttribute subAttribute =. new SimpleAttribute(a,c);
           return (SimpleAttribute) DefaultAttributeFactory.createAttribute(b, subAttribute);
        }
```

File: modules/charon-core/src/main/java/org/wso2/charon3/core/objects/RoleV2.java [hunk 2]
```diff
@@ -0,0 +1,621 @@
...
+                        displayNames.add((String) ((SimpleAttribute) (subAttributesList
+                                .get(SCIMConstants.CommonSchemaConstants.DISPLAY))).getValue());
+                    }
+
+                }
+                return displayNames;
+            }
+        }
+
+        return displayNames;
```
Comment (thanujalk) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/objects/RoleV2.java:149]: Let's avoid methods getting expanded to the right. let's return from the reverse if conditions.

File: modules/charon-core/src/main/java/org/wso2/charon3/core/objects/Group.java (hunk 1 above)
Comment (AnuradhaSK) [reply modules/charon-core/src/main/java/org/wso2/charon3/core/objects/Group.java]: fixed in [ca64654](https://github.com/wso2/charon/pull/392/commits/ca64654c3f9c03eb95c521cc2f548d9206cbdb4e)

File: modules/charon-core/src/main/java/org/wso2/charon3/core/objects/RoleV2.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (AnuradhaSK) [reply modules/charon-core/src/main/java/org/wso2/charon3/core/objects/RoleV2.java]: fixed in [ca64654](https://github.com/wso2/charon/pull/392/commits/ca64654c3f9c03eb95c521cc2f548d9206cbdb4e)

Comment (AnuradhaSK) [reply modules/charon-core/src/main/java/org/wso2/charon3/core/objects/RoleV2.java]: fixed in [ca64654](https://github.com/wso2/charon/pull/392/commits/ca64654c3f9c03eb95c521cc2f548d9206cbdb4e)fixed in [ca64654](https://github.com/wso2/charon/pull/392/commits/ca64654c3f9c03eb95c521cc2f548d9206cbdb4e)

File: modules/charon-core/src/main/java/org/wso2/charon3/core/protocol/endpoints/RoleResourceV2Manager.java [hunk 3]
```diff
@@ -0,0 +1,788 @@
...
+
+            // Retrieve the old object.
+            RoleV2 oldRole = roleManager.getRole(id, requestAttributes);
+            if (oldRole != null) {
+                RoleV2 newRole = (RoleV2) ServerSideValidator.validateUpdatedSCIMObject(oldRole, role, schema);
+                updatedRole = roleManager.updateRole(oldRole, newRole);
+            } else {
+                String error = "No role exists with the given id: " + id;
+                throw new NotFoundException(error);
+            }
```
Comment (AnuradhaSK) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/protocol/endpoints/RoleResourceV2Manager.java:270]: reserve the logic and remove else

File: modules/charon-core/src/main/java/org/wso2/charon3/core/objects/Group.java [hunk 4]
```diff
@@ -411,4 +419,89 @@ private void setRole(ComplexAttribute groupPropertiesAttribute) throws CharonExc
...
+
+    private SimpleAttribute getSimpleAttribute(String attributeName, String attributeValue,
+                                               SCIMAttributeSchema attributeSchema)
+            throws CharonException, BadRequestException {
+
+        return (SimpleAttribute) DefaultAttributeFactory.createAttribute(attributeSchema,
+                new SimpleAttribute(attributeName, attributeValue));
+    }
+
+    private void setRoleV2(ComplexAttribute groupPropertiesAttribute) throws CharonException, BadRequestException {
```
Comment (AnuradhaSK) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/objects/Group.java:493]: this should be rolePropertiesAttribute

File: modules/charon-core/src/main/java/org/wso2/charon3/core/objects/RoleV2.java [hunk 5]
```diff
@@ -0,0 +1,648 @@
...
+        for (Attribute subValue : subValuesList) {
+            ComplexAttribute complexAttribute = (ComplexAttribute) subValue;
+            Map<String, Attribute> subAttributesList = complexAttribute.getSubAttributesList();
+            if (subAttributesList == null ||
+                    !subAttributesList.containsKey(SCIMConstants.CommonSchemaConstants.VALUE)) {
+                continue;
+            }
+            permissionValuesList.add((String) ((SimpleAttribute) (subAttributesList
+                    .get(SCIMConstants.CommonSchemaConstants.VALUE))).getValue());
+
```
Comment (AnuradhaSK) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/objects/RoleV2.java:388]:
```suggestion
```

File: modules/charon-core/src/main/java/org/wso2/charon3/core/objects/RoleV2.java [hunk 6]
```diff
@@ -0,0 +1,648 @@
...
+        if (!this.isAttributeExist(SCIMConstants.RoleSchemaConstants.PERMISSIONS)) {
+            return Collections.emptyList();
+        }
+        MultiValuedAttribute permissions = (MultiValuedAttribute) this.attributeList
+                .get(SCIMConstants.RoleSchemaConstants.PERMISSIONS);
+        List<Attribute> values = permissions.getAttributeValues();
+        List<String> permissionDisplayNames = new ArrayList<>();
+        if (values == null) {
+            return Collections.emptyList();
+
```
Comment (AnuradhaSK) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/objects/RoleV2.java:409]:
```suggestion
```

File: modules/charon-core/src/main/java/org/wso2/charon3/core/schema/SCIMConstants.java [hunk 7]
```diff
@@ -524,24 +541,44 @@ public static class RoleSchemaConstants {
...
+        public static final String APPLICATION = "application";
+        public static final String ORGANIZATION = "organization";
+
         // Attributes descriptions of the attributes found in Role Schema.
         public static final String DISPLAY_NAME_DESC = "A human-readable name for the Role. REQUIRED.";
         public static final String USERS_DESC = "A list of users of the role.";
         public static final String GROUPS_DESC = "A list of groups of the role.";
         public static final String PERMISSIONS_DESC = "A list of permissions of the role.";
+        public static final String ASC_APPLICATIONS_DESC = "A list of associated applications of the role.";
+        public static final String AUDIENCE_DESC = "The role usable scope.";
```
Comment (AnuradhaSK) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/schema/SCIMConstants.java:563]:
```suggestion
        public static final String AUDIENCE_DESC = "The role usable audience.";
```

File: modules/charon-core/src/main/java/org/wso2/charon3/core/schema/SCIMResourceSchemaManager.java [hunk 8]
```diff
@@ -215,6 +217,10 @@ public SCIMResourceTypeSchema getRoleResourceSchema() {
         return SCIMSchemaDefinitions.SCIM_ROLE_SCHEMA;
     }
 
+    public SCIMResourceTypeSchema getRoleResourceV2Schema() {
+
+        return SCIMSchemaDefinitions.SCIM_ROLE_V2_SCHEMA;
+    }
```
Comment (AnuradhaSK) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/schema/SCIMResourceSchemaManager.java:223]: leave a new line

File: modules/charon-core/src/main/java/org/wso2/charon3/core/schema/SCIMSchemaDefinitions.java [hunk 9]
```diff
@@ -513,6 +513,26 @@ public static class SCIMUserSchemaDefinition {
                         SCIMDefinitions.Uniqueness.NONE, null, new ArrayList<>
                                 (Arrays.asList(SCIMDefinitions.ReferenceType.USER, SCIMDefinitions.ReferenceType
                                         .ROLE)), null);
+        public static final SCIMAttributeSchema ROLES_AUDIENCE_VALUE = SCIMAttributeSchema.createSCIMAttributeSchema(
```
Comment (AnuradhaSK) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/schema/SCIMSchemaDefinitions.java:516]: leave a line

**Comment by AnuradhaSK** [comment]
Track to move previous RoleV2 out of Charon from: https://github.com/wso2/product-is/issues/17096

File: modules/charon-core/src/main/java/org/wso2/charon3/core/schema/SCIMConstants.java [hunk 10]
```diff
@@ -102,6 +105,11 @@ public static class CommonSchemaConstants {
         public static final String REF = "$ref";
         public static final String VALUE = "value";
 
+        public static final String AUDIENCE_VALUE = "audienceValue";
+        public static final String AUDIENCE_DISPLAY = "audienceDisplay";
+        public static final String AUDIENCE_TYPE = "audienceType";
+
+
```
Comment (shashimalcse) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/schema/SCIMConstants.java:112]:
```suggestion

```

Comment (AnuradhaSK) [reply modules/charon-core/src/main/java/org/wso2/charon3/core/schema/SCIMConstants.java]: fixed in ccf0c9f

File: modules/charon-core/src/main/java/org/wso2/charon3/core/objects/RoleV2.java [hunk 11]
```diff
@@ -0,0 +1,649 @@
...
+            ComplexAttribute complexAttribute = (ComplexAttribute) subValue;
+            Map<String, Attribute> subAttributesList = complexAttribute.getSubAttributesList();
+            if (subAttributesList == null ||
+                    !subAttributesList.containsKey(SCIMConstants.CommonSchemaConstants.DISPLAY)) {
+                continue;
+            }
+            displayNames.add((String) ((SimpleAttribute) (subAttributesList
+                    .get(SCIMConstants.CommonSchemaConstants.DISPLAY))).getValue());
+        }
+        return displayNames;
```
Comment (thanujalk) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/objects/RoleV2.java:150]: Looks like the same logic is repeated in multiple get methods. Better write a private method that takes RoleSchemaConstants.X as input and returns the list.

File: modules/charon-core/src/main/java/org/wso2/charon3/core/objects/RoleV2.java [hunk 12]
```diff
@@ -0,0 +1,649 @@
...
+     * Get the audience type of the role.
+     */
+    public String getAudienceType() {
+
+        SCIMAttributeSchema audienceDefinition = SCIMSchemaDefinitions.SCIMRoleSchemaDefinition.AUDIENCE;
+        SCIMAttributeSchema audienceTypeDefinition = SCIMSchemaDefinitions.SCIMRoleSchemaDefinition.AUDIENCE_TYPE;
+        return getComplexAttribute(audienceDefinition).map(aud -> getSimpleAttribute(audienceTypeDefinition, aud)
+                .map(rethrowFunction(SimpleAttribute::getStringValue))
+                .orElse(null)).orElse(null);
+    }
```
Comment (thanujalk) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/objects/RoleV2.java:576]: can extract a private method

File: modules/charon-core/src/main/java/org/wso2/charon3/core/objects/User.java [hunk 13]
```diff
@@ -915,6 +917,96 @@ private void setRole(ComplexAttribute groupPropertiesAttribute) throws CharonExc
...
+     * Set the assigned V2 roles of the user.
+     *
+     * @param role RoleV2 object.
+     * @throws CharonException     CharonException.
+     * @throws BadRequestException BadRequestException.
+     */
+    public void setRoleV2(RoleV2 role) throws CharonException, BadRequestException {
+
+        SimpleAttribute valueSimpleAttribute = null;
+        SimpleAttribute displaySimpleAttribute;
```
Comment (thanujalk) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/objects/User.java:930]: no usage

Comment (thanujalk) [reply modules/charon-core/src/main/java/org/wso2/charon3/core/objects/User.java:930]: same for the below ones

Comment (AnuradhaSK) [reply modules/charon-core/src/main/java/org/wso2/charon3/core/objects/User.java]: fixed in [25b26c2](https://github.com/wso2/charon/pull/392/commits/25b26c2228a54f30201b37c15e189987e55ccf32)

File: modules/charon-core/src/main/java/org/wso2/charon3/core/objects/RoleV2.java (hunk 12 above)
Comment (AnuradhaSK) [reply modules/charon-core/src/main/java/org/wso2/charon3/core/objects/RoleV2.java]: fixed in [25b26c2](https://github.com/wso2/charon/pull/392/commits/25b26c2228a54f30201b37c15e189987e55ccf32)

File: modules/charon-core/src/main/java/org/wso2/charon3/core/objects/RoleV2.java (hunk 11 above)
Comment (AnuradhaSK) [reply modules/charon-core/src/main/java/org/wso2/charon3/core/objects/RoleV2.java]: fixed in [25b26c2](https://github.com/wso2/charon/pull/392/commits/25b26c2228a54f30201b37c15e189987e55ccf32)

**Review by thanujalk — APPROVED** [review]

**Pull request merged**